job-info & job-manager& flux-jobs: Support job annotations (annotations part 3)

chu11 opened the pull request:
Built on top of PR #3062, this PR completes phase 1 job annotation support in `flux-core`.

- job-manager - support a new annotations event, which will publish annotations

- job-info - read in annotations events from the above for querying by job-list services

- job-manager - support a new user annotation service, allowing users to annotate their jobs
  - add command support in `flux job annotate`
  - is the user interface ok?  I supported a simple `flux job annotate id key value` syntax, and `value` can be `-` for stdin.
  - by default, everything is put under the `user` object in the annotations.

- flux-jobs - support listing any annotations
  - the annotation fields listed by the user can be arbitrary.  `flux-jobs` has a rule that any annotation that doesn't exist results in an empty string.  So this is even if the user inputs bogus annotations.  The reason we have to do this is b/c there is no way to know what annotations  are legal or illegal, given any scheduler or user can come up with any annotations they want. 
  - do we want to support a few special case annotations such as `sched` and/or `user`?  So the user doesn't have to type "annotations." for each of them?
  - the output headers ok?  I chose to upper case everything after "annotations." as the header.

An example output for:

```
jobid=`flux mini submit sleep 1000`                                                                                                         
sleep 5                                                                                                                                     
flux job annotate $jobid foo 1234                                                                                                           
flux jobs -n --format="{id} --- {annotations}"                                                                                              
flux jobs -n --format="{id} --- {annotations.sched}"                                                                                        
flux jobs -n --format="{id} --- {annotations.user}"                                                                                         
flux jobs -n --format="{id} --- {annotations.sched.resource_summary}"                                                                       
flux jobs -n --format="{id} --- {annotations.sched.foobar}"                                                                                 
flux jobs -n --format="{id} --- {annotations.user.foo}"                                                                                     
flux jobs -n --format="{id} --- {annotations.user.a.illegal.key}"  
```

```
>src/cmd/flux start ./submit_annotate.sh
15804137472 --- {"sched": {"resource_summary": "rank0/core0"}, "user": {"foo": 1234}}
15804137472 --- {"resource_summary": "rank0/core0"}
15804137472 --- {"foo": 1234}
15804137472 --- rank0/core0
15804137472 --- 
15804137472 --- 1234
15804137472 --- 
```

example w/ output headers (sorta ugly for this example)

```
JOBID --- ANNOTATIONS
15854469120 --- {"sched": {"resource_summary": "rank0/core0"}, "user": {"foo": 1234}}
JOBID --- SCHED
15854469120 --- {"resource_summary": "rank0/core0"}
JOBID --- USER
15854469120 --- {"foo": 1234}
JOBID --- SCHED.RESOURCE_SUMMARY
15854469120 --- rank0/core0
JOBID --- SCHED.FOOBAR
15854469120 --- 
JOBID --- USER.FOO
15854469120 --- 1234
JOBID --- USER.A.ILLEGAL.KEY
15854469120 --- 
```

grondo commented:
> do we want to support a few special case annotations such as sched and/or user? So the user doesn't have to type "annotations." for each of them?

That might be a good idea. We could also have meaningful default headings for those cases.

chu11 commented:
> We could also have meaningful default headings for those cases.

Have anything in mind?  I could see a couple of special cases, such as `T_ESTIMATE`, "REASON PENDING" and perhaps "RESOURCE SUMMARY", per RFC27 pre-defined fields (i would probably just drop the `sched.` and remove the underscore).  But those are the only ones that come to mind.

chu11 commented:
rebased on master & fixed up some conflicts given the unit tests added in PR #3062

grondo commented:
> Have anything in mind? I could see a couple of special cases, such as T_ESTIMATE, "REASON PENDING" and perhaps "RESOURCE SUMMARY", per RFC27 pre-defined fields (i would probably just drop the sched. and remove the underscore). 

Yeah, `T_ESTIMATE` and `REASON` were the two I had in mind. `REASON` will probably end up in default output of `flux jobs` I'd think.

chu11 commented:
- Re-pushed with a few changes

  - added some more tests to t2205-job-manager-annotate to handle a user annotating a job with a sub-object.

  - removed some tests in t2205-job-manager-annotate that were excessive

  - added the field names "sched" and "user" to be short hand for  `annotations.sched` and `annotations.user`.

  - per discussion above, made the following RFC27 annotations have   specific header outputs

```
+        # The following are special pre-defined cases per RFC27
+        "annotations.sched.t_estimate": "T_ESTIMATE",
+        "annotations.sched.reason_pending": "REASON",
+        "annotations.sched.resource_summary": "RESOURCES",
+        "sched": "SCHED",
+        "sched.t_estimate": "T_ESTIMATE",
+        "sched.reason_pending": "REASON",
+        "sched.resource_summary": "RESOURCES",
+        "user": "USER",
```

  - fixed some typos and English documentation errors I found

Went ahead and squashed since my tweaks were a bit jumbled up, I hope thats ok :P

garlick commented:
What am I doing wrong here?
```
$ flux mini run hostname
sun1
$ flux jobs list -a
             JOBID USER     NAME       STATUS NTASKS NNODES  RUNTIME RANKS
       70514638848 garlick  hostname       CD      1      1   0.079s 0
$ flux job annotate 70514638848 fubar xyz
flux-job: invalid value: invalid token near 'xyz': Success
$ flux job annotate 70514638848 user.fubar xyz
flux-job: invalid value: invalid token near 'xyz': Success
$ flux job annotate 70514638848 user.fubar 42
flux-job: flux_rpc_get: No such file or directory
```
Error messages are not helping me figure this out :-)

chu11 commented:
The first two error messages are b/c "xyz" isn't legal json.  Should be `\"xyz\"`.  But I should probably just assume its a string if its not legal json?

The latter is b/c you're trying to annotate a completed job.  I should definitely put in a better error message in that case.

garlick commented:
Thanks, sorry, first naive experience!  Note that jansson functions don't set errno, so the No error is because log_err_exit() is being used to log errors.  Better messages would be good.

chu11 commented:
> Thanks, sorry, first naive experience! 

My similarly naive implementation :-)  I implemented the quickest thing to try things out, forgot to go back and make the error messages better :P

> Edit: looks like the service sends back useful string error messages, but flux jobs annotate doesn't display them. Maybe use future_strerror(f, errno) macro there?

Realized it too :-)

Re-pushed with these fixes:

```
>flux job annotate 81872814080 foo bar
flux-job: invalid json value: invalid token near 'bar'
>flux job annotate 81872814080 foo 1234
flux-job: annotate: unknown job id
```

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1364,6 +1437,19 @@ static struct job *eventlog_restart_parse (struct info_ctx *ctx,
+                if (json_unpack (context,
+                                 "{ s:o }",
+                                 "annotations", &annotations) < 0) {
+                    flux_log_error (ctx->h, "%s: alloc context for %ju invalid",
```
errno not set by `json_unpack()` but `flux_log_error()` is called.

also would json_object_get() be a lighter weight equivalent to `json_unpack ("s:o")`?

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -211,6 +219,8 @@ struct event_batch *event_batch_create (struct event *event)
     if (!(batch->state_trans = json_array ()))
         goto nomem;
+    if (!(batch->annotations = json_array ()))
+        goto nomem;
```
One thought would be to leave annotations and state_trans NULL here and create on demand, since a `struct event batch` seems likely to contain only one of the two, unless the system is very busy.

garlick commented on src/cmd/flux-job.c:
```diff
@@ -2854,6 +2862,62 @@ int cmd_wait (optparse_t *p, int argc, char **argv)
+        value = (const char *)valbuf;
+    }
+    else
+        value = argv[optindex + 2];
```
Should exactly argc==3 be required if all remaining args aren't being combined to construct the value?

Also, it would be sort of handy if one could specify a string value without the quotes.

chu11 commented on src/modules/job-manager/annotate.c:
```diff
@@ -0,0 +1,204 @@
+        errno = ENOENT;
+        goto error;
+    }
+    if (flux_msg_cred_authorize (cred, job->userid) < 0) {
```
credit goes to you, it was in the priority & raise code :-)

chu11 commented on src/modules/job-info/job_state.c:
```diff
@@ -1237,6 +1243,73 @@ void job_state_cb (flux_t *h, flux_msg_handler_t *mh,
+        json_t *o;
+        flux_jobid_t id;
+
+        if (!json_is_array (value)) {
```
good idea, and I can duuplicate this cleanup in `update_jobs()` for the job state transitions

chu11 commented on src/modules/job-manager/annotate.c:
```diff
@@ -0,0 +1,204 @@
+    flux_msg_handler_t **handlers;
+};
+
+void annotations_clear (struct job *job, bool *cleared)
```
definitely could, but I think this is more clear.  Minimally `annotations_clear()` is called in `update_recursive()`, so we'd be making a recursive call just to clear the data, which seems unnecessary.

chu11 commented on src/modules/job-manager/annotate.c:
```diff
@@ -0,0 +1,204 @@
+/* we want to delete items set to 'null', so this is not the same
+ * as json_object_update_recursive() in jansson 2.13.1
+ */
+void update_recursive (struct job *job, json_t *orig, json_t *new,
```
I picked `update_annotations_recursive()`. Since its annotations it should be clear we're following RFC27 semantics.

chu11 commented:
rebased on master & re-pushed addressing the stuff @garlick mentioend above.  There's a lot of fixups and a new cleanup commit that I should re-order in the commit series.  Would it be easier to review by just squashing everything?

garlick commented:
Thanks!  Squashing would probably be preferable.

chu11 commented:
> Thanks! Squashing would probably be preferable.

Squashed and fixed a few nits along the way.  Rebased while I was at it.

garlick commented:
How about one more rebase to resolve the conflict and then I'll do another review pass?

chu11 commented:
> How about one more rebase to resolve the conflict and then I'll do another review pass?

Already working on it :-) hit #3071

chu11 commented:
rebased and re-pushed, added a fix for #3074 on top.  It reverts a testsuite performance enhancement that @trws did.  Since its only performance related, should be fine to revert.

chu11 commented:
rebased and repushed.  I removed the #3074 fix since that won't affect travis.  I'll put that fix in a separate PR.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1238,6 +1244,80 @@ void job_state_cb (flux_t *h, flux_msg_handler_t *mh,
+    json_t *value;
+
+    if (!json_is_array (annotations)) {
+        flux_log (ctx->h, LOG_ERR, "%s: annotations EPROTO", __FUNCTION__);
```
Improve error message?  Maybe "annotations event is not an array" and skip the __FUNCTION__?

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1238,6 +1244,80 @@ void job_state_cb (flux_t *h, flux_msg_handler_t *mh,
+        flux_jobid_t id;
+        json_t *aValue;
+
+        if (parse_annotation (value, &id, &aValue) < 0) {
```
How about "job %ju annotation parse error"

chu11 replied:
but I won't know the ID :-), i'll adjust to "annotation parse error", to differentiate from the error above it

garlick commented on src/cmd/flux-job.c:
```diff
@@ -2827,6 +2835,64 @@ int cmd_wait (optparse_t *p, int argc, char **argv)
+    if (!(h = flux_open (NULL, 0)))
+        log_err_exit ("flux_open");
+
+    id = parse_arg_unsigned (argv[optindex], "id");
```
job ID's should now be parsed with `parse_jobid()` to support F58 and other representations.

chu11 commented:
> So the annotations added with flux job annotate are not sticky across a restart. Is that intended? Should there be an annotate event in the job eventlog that is recovered by the job manager?

If the job has run (i.e. alloc successful), then the user annotation should be sticky.  We defined it as non-sticky with a job that hasn't run.  I'll double check that it works and we have a test for it.

chu11 commented:
rebased & re-pushed with fixes for the above issues

garlick commented:
The test would need to reload the job manager as well, but I take your point - a user annotation will become part of the alloc event context once the job is allocated, along with the scheduler annotations.  The problem is that the scheduler might be expected to re-annotate pending jobs after a reload, but user annotations would not be restored in that way.

I think it would be surprising if I were to annotate my job and have the annotations disappear after a system maintenance window, for example.  I'll open an issue on that.  It seems like we might want to take a step back and re-evaluate how annotations are made persistent.  (If they are now a general concept, do they belong in the alloc eventlog entry, for example?)

No need to hold up this PR on that though, IMHO.

chu11 commented:
>  I'll open an issue on that. It seems like we might want to take a step back and re-evaluate how annotations are made persistent. (If they are now a general concept, do they belong in the alloc eventlog entry, for example?)

Sounds good.  We'll have to update RFC27 as well.  The key points in there:

```
The job manager posts an alloc event in response to the successful allocation of resources. A snapshot of job's annotation dictionary, after the above update, is included in the alloc event context per RFC 21, thus preserving it in job record when the allocation is successful.
```

```
Annotations SHALL be considered volatile until a SUCCESS response is received to the sched.alloc request, as described in Alloc Success above. Annotations SHALL be discarded by the job manager if the allocation fails.
```

chu11 commented:
> No need to hold up this PR on that though, IMHO. 

Shall I squash?

garlick commented:
sure!

chu11 commented:
squashed and re-pushed

grondo commented:
I get a reproducible failure in `t2205-job-manager-annotate.t` when the test is run with `FLUX_TEST_VALGRIND` set:
```
ok 26 - job-manager: cancel all jobs
not ok 27 - job-manager: no annotations (IIII)
#	
#	        jmgr_check_no_annotations $(cat job1.id) &&
#	        jmgr_check_no_annotations $(cat job2.id) &&
#	        jmgr_check_no_annotations $(cat job3.id) &&
#	        jmgr_check_no_annotations $(cat job4.id) &&
#	        jmgr_check_no_annotations $(cat job5.id)
#	
ok 28 - job-manager: no annotations in job-info (IIII)
ok 29 - job-manager: no annotations in job-info (IIII)
ok 30 - job-manager: remove sched-dummy
ok 31 - job-manager: remove job-manager, job-ingest
# failed 1 among 31 test(s)
1..31
```

Maybe some synchronization needed between the cancelation of jobs and the next test?

grondo commented on t/t2204-job-manager-dummysched-unlimited.t:
```diff
@@ -117,10 +167,38 @@ test_expect_success HAVE_JQ 'job-manager: annotate job id 4 (RIRSI)' '
+# running jobs free resources.
 test_expect_success 'job-manager: cancel all jobs' '
-        flux job cancel $(cat job1.id) &&
+        flux job cancel $(cat job4.id) &&
         flux job cancel $(cat job3.id) &&
-        flux job cancel $(cat job4.id)
+        flux job cancel $(cat job1.id)
```
Just a suggestion, I think `flux job cancelall` takes a `--states` argument, so this test could be shortened and made self documenting with something like (untested):
```
        flux job cancelall --states=SCHED -f &&
        flux job cancelall -f
```

chu11 commented:
> Maybe some synchronization needed between the cancelation of jobs and the next test?

I think you're right.  I removed the "job states" tests b/c I thought them superfluous, but they are probably necessary for synchronization.

grondo commented:
I'm going to push @chu11's latest work here. All I did was run `git rebase -i --autosquash upstream/master`.

garlick commented:
Setting MWP.  Thanks!
